flight/PiOS/Common/pios_ms5xxx.c

```diff
@@ -0,0 +1,700 @@
+static void PIOS_MS5XXX_Task(void *parameters);
+
+/* Private types */
+//! The valid hardware buses
+enum PIOS_BUS_TYPE {
+	PIOS_SPI_BUS,
+	PIOS_I2C_BUS,
+};
+
+/* Local Types */
+
+enum PIOS_MS5XXX_DEV_MAGIC {
+	PIOS_MS5XXX_DEV_MAGIC = 0x1c50bcf2, // md5 of `PIOS_MS5XXX_DEV_MAGIC`
+};
+
+enum conversion_type {
+	PRESSURE_CONV,
+	TEMPERATURE_CONV
+};
+
+struct ms5xxx_dev {
+	const struct pios_ms5xxx_cfg * cfg;
+	enum PIOS_BUS_TYPE pios_bus_type; // Bus type can be either SPI or I2C
+	uint32_t dev_id; // Valid for both SPI and I2C
+	uint32_t slave_num; // Only used for SPI bus chip select
+	uint8_t ms5xxx_i2c_addr; // Only used for I2C 7-bit address
+	struct pios_thread *task;
+	struct pios_queue *queue;
+
+	int64_t pressure;
+	int64_t temperature_unscaled;
+	uint16_t calibration[6];
+	enum conversion_type current_conversion_type;
+	enum PIOS_MS5XXX_DEV_MAGIC magic;
+
+	struct pios_semaphore *busy;
+};
+
+static struct ms5xxx_dev *dev;
+
+/**
+ * @brief Allocate a new device
+ */
+static struct ms5xxx_dev * PIOS_MS5XXX_alloc(void)
+{
+	struct ms5xxx_dev *ms5xxx_dev;
+
+	ms5xxx_dev = (struct ms5xxx_dev *)PIOS_malloc(sizeof(*ms5xxx_dev));
+	if (!ms5xxx_dev) {
+		return (NULL);
+	}
+
+	memset(ms5xxx_dev, 0, sizeof(*ms5xxx_dev));
+
+	ms5xxx_dev->queue = PIOS_Queue_Create(1, sizeof(struct pios_sensor_baro_data));
+	if (ms5xxx_dev->queue == NULL) {
+		PIOS_free(ms5xxx_dev);
+		return NULL;
+	}
+
+	ms5xxx_dev->magic = PIOS_MS5XXX_DEV_MAGIC;
+
+	ms5xxx_dev->busy = PIOS_Semaphore_Create();
+	PIOS_Assert(ms5xxx_dev->busy != NULL);
+
+	return ms5xxx_dev;
+}
+
+/**
+ * @brief Validate the handle to the device
+ * @returns 0 for valid device or <0 otherwise
+ */
+static int32_t PIOS_MS5XXX_Validate(struct ms5xxx_dev *dev)
+{
+	if (dev == NULL)
+		return -1;
+	if (dev->magic != PIOS_MS5XXX_DEV_MAGIC)
+		return -2;
+	if (dev->dev_id == 0)
+		return -3;
+	return 0;
+}
+
+
+/**
+ * @brief PIOS_MS5XXX_Init Initializes the MS5xxx chip
+ * @param cfg Configuration structure. Note that it is not deep copied.
+ * @return 0 on success. -1 if the allocation fails, -2 if the board
+ * setup file does not define the model of chip, and -3 if the write
+ * command fails.
+ */
+int32_t PIOS_MS5XXX_Init(const struct pios_ms5xxx_cfg *cfg)
+{
+	// Set the pointer to the configuration struct. Note that it is not deep copied.
+	dev->cfg = cfg;
+
+	// Check that a model of the chip was chosen
+	if (cfg->pios_ms5xxx_model == 0) {
+		return -2;
+	}
+
+	if (PIOS_MS5XXX_WriteCommand(MS5XXX_RESET) != 0) {
+		return -3;
+	}
+
+	PIOS_DELAY_WaitmS(20);
+
+	uint8_t data[2];
+
+	/* Calibration parameters */
+	for (int i = 0; i < NELEMENTS(dev->calibration); i++) {
+		PIOS_MS5XXX_Read(MS5XXX_CALIB_ADDR + i * 2, data, 2);
+		dev->calibration[i] = (data[0] << 8) | data[1];
+	}
+
+	PIOS_SENSORS_Register(PIOS_SENSOR_BARO, dev->queue);
+
+	dev->task = PIOS_Thread_Create(
+			PIOS_MS5XXX_Task, "pios_ms5xxx", MS5XXX_TASK_STACK_BYTES, NULL, MS5XXX_TASK_PRIORITY);
```

Comment:
Should this task be added to the task manager UAVObject? I don't really see any downsides to this, but it wasn't done before so I didn't add it.